openmc/deplete/batchwise.py

```diff
@@ -0,0 +1,1316 @@
+        The parametrization can either be at constant volume or constant
+        density, according to user input.
+        Default is constant volume.
+
+        Parameters
+        ----------
+        param : float
+            Model function variable, total grams of material to add or remove
+
+        Returns
+        -------
+        model :  openmc.model.Model
+            Openmc parametric model
+        """
+        for rank in range(comm.size):
+            number_i = comm.bcast(self.operator.number, root=rank)
+
+            for mat in number_i.materials:
+                nuclides = []
+                densities = []
+
+                if int(mat) == self.material.id:
+
+                    if self.density_treatment == "constant-density":
+                        vol = number_i.get_mat_volume(mat) + (
+                            param / self.material.get_mass_density()
+                        )
+
+                    elif self.density_treatment == "constant-volume":
+                        vol = number_i.get_mat_volume(mat)
+
+                    for nuc in number_i.index_nuc:
+                        # check only nuclides with cross sections data
+                        if nuc in self.operator.nuclides_with_data:
+                            if nuc in self.mat_vector:
+                                # units [#atoms/cm-b]
+                                val = 1.0e-24 * (
+                                    number_i.get_atom_density(mat, nuc)
+                                    + param
+                                    / atomic_mass(nuc)
+                                    * AVOGADRO
+                                    * self.mat_vector[nuc]
+                                    / vol
+                                )
+
+                            else:
+                                # get normalized atoms density in [atoms/b-cm]
+                                val = 1.0e-24 * number_i[mat, nuc] / vol
+
+                            if val > 0.0:
+                                nuclides.append(nuc)
+                                densities.append(val)
+
+                else:
+                    # for all other materials, still check atom density limits
+                    for nuc in number_i.nuclides:
+                        if nuc in self.operator.nuclides_with_data:
+                            # get normalized atoms density in [atoms/b-cm]
+                            val = 1.0e-24 * number_i.get_atom_density(mat, nuc)
+
+                            if val > 0.0:
+                                nuclides.append(nuc)
+                                densities.append(val)
+
+                # set nuclides and densities to the in-memory model
+                openmc.lib.materials[int(mat)].set_densities(nuclides, densities)
+
+        # always need to return a model
+        return self.model
+
+    def _calculate_volumes(self, res):
+        """Uses :meth:`openmc.batchwise._search_for_keff` solution as grams of
```

Comment:
Please ensure docstrings have a single line and then an empty line before new content